### ISSUE TITLE
Correctly uploaded file was removed during multipart upload

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#903]
I have a problem that occurred on production cluster and resulted in loss of an object that wasn't noticed by application.
The application logic is as follows: there is a directory full of source objects: O1, O2, O3 etc. New objects are constantly appearing there, they always have a unique name.  Around 20 applications check this directory all the time (all at once) and pick objects to process, if the objects are correct, they extract destination objects from these source objects, like this: O1 -> D1, D2, .. Dn. Different instances never touch the same source object but the same destination object can arrive from different source objects, e.g. O1-> .. D5 .., and O10 -> ... D5 ... . All destination objects are saved to LeoFS with this logic: first, an SQL database is queried to check if we already have that object. If we do, it's thrown away; if we don't, it's saved to storage - right now LeoFS (in unconditional PUT operation). After which we mark object as present in the database. Of course, it can be marked as present already if another instance was saving the same object (coming from different source object) as we were querying the database, then by the time we're done saving it's already uploaded by that instance as well and the presence mark is set. Which means we have overwritten the object that the other instance was saving, but it's not a big deal since both copies are the same. Old storage implemented atomic PUT operation (by writing temporary object and doing rename in the end), so in case of conflict data couldn't break, and it did not matter how the conflict was resolved as all copies were the same.
When all destination objects are uploaded or skipped without a problem, application removes source object and goes to the next one. If any error occurs, application instance crashes and restarts and picks the same object again (or another instance would've picked it by then).

The problem that happened is as follows. Five application instances extract an object d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000 from different source container, compress it and try to upload to storage (all at the same time, around 22:03 - 22:04). The size of compressed object is 6.2M so we upload it as multipart object through boto3 (chunk size 5M). Three instances succeed and mark object as present in database, two log the same error:
```
<class 'botocore.exceptions.ClientError'>: 'An error occurred (ServiceUnavailable) when calling the CompleteMultipartUpload operation (reached max retries: 1): Please reduce your request rate.
```
Naturally, two source containers that contained this object, during processing of which the error happened are retried but the object was already marked as present at this time so it was just skipped.

The problem: object is actually deleted now. No one did any deletes (of this or any other objects), the only deletes probably could've been issued by gateway as part of cleanup operation.
```
[vm@bodies-master ~]$ leofs-adm whereis 'body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz'
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
 del?  |                  node                 |             ring address             |    size    |   checksum   |  has children  |  total chunks  |     clock      |             when            
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
  *    | bodies@stor03.selectel.cloud.lan      | 9c8160ba024dfa21ea8ef46396a005ad     |         0B |   d41d8cd98f | false          |              0 | 55c4f9988d898  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor04.selectel.cloud.lan      | 9c8160ba024dfa21ea8ef46396a005ad     |         0B |   d41d8cd98f | false          |              0 | 55c4f9988d898  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor06.selectel.cloud.lan      | 9c8160ba024dfa21ea8ef46396a005ad     |         0B |   d41d8cd98f | false          |              0 | 55c4f9988d898  | 2017-10-24 22:04:25 +0300
```
No logs on gateway. There are strange logs on storage nodes. Both info and error logs here. On stor01:
```
[I]	bodies@stor01.selectel.cloud.lan	2017-10-24 22:03:51.612798 +0300	1508871831	leo_storage_handler_del_directory:handle_call/3 - enqueue	149	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[E]	bodies@stor01.selectel.cloud.lan	2017-10-24 22:03:51.892663 +0300	1508871831	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n06d9c70d867d6ba1b14cbad1f4df775d">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-10-24 22:03:52.697662 +0300	1508871832	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n06d9c70d867d6ba1b14cbad1f4df775d">>},{cause,not_found}]
[I]	bodies@stor01.selectel.cloud.lan	2017-10-24 22:03:56.229110 +0300	1508871836	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```
On stor02:
```
[E]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:50.970029 +0300	1508871830	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n44688f514c7dc8ec666e5fd3b18879db">>},{cause,not_found}]
[E]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:51.593879 +0300	1508871831	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n44688f514c7dc8ec666e5fd3b18879db">>},{cause,not_found}]
[E]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:51.610770 +0300	1508871831	leo_storage_handler_object:put/4	424	[{from,storage},{method,delete},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>},{req_id,14911882},{cause,not_found}]
[I]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:51.613119 +0300	1508871831	leo_storage_handler_del_directory:handle_call/3 - enqueue	149	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[E]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:52.710940 +0300	1508871832	leo_storage_handler_object:put/4	424	[{from,storage},{method,delete},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>},{req_id,73489324},{cause,not_found}]
[I]	bodies@stor02.selectel.cloud.lan	2017-10-24 22:03:56.951100 +0300	1508871836	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```
On stor03:
```
[E]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:50.968994 +0300	1508871830	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n44688f514c7dc8ec666e5fd3b18879db">>},{cause,not_found}]
[E]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:51.592972 +0300	1508871831	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n44688f514c7dc8ec666e5fd3b18879db">>},{cause,not_found}]
[I]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:51.613856 +0300	1508871831	leo_storage_handler_del_directory:run/5	143	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[E]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:51.891844 +0300	1508871831	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n06d9c70d867d6ba1b14cbad1f4df775d">>},{cause,not_found}]
[E]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:52.696618 +0300	1508871832	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n06d9c70d867d6ba1b14cbad1f4df775d">>},{cause,not_found}]
[I]	bodies@stor03.selectel.cloud.lan	2017-10-24 22:03:54.841349 +0300	1508871834	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```

On stor04:
```
[I]	bodies@stor04.selectel.cloud.lan	2017-10-24 22:03:51.612658 +0300	1508871831	leo_storage_handler_del_directory:handle_call/3 - enqueue	149	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[I]	bodies@stor04.selectel.cloud.lan	2017-10-24 22:03:55.922602 +0300	1508871835	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```

On stor05:
```
[I]	bodies@stor05.selectel.cloud.lan	2017-10-24 22:03:51.612644 +0300	1508871831	leo_storage_handler_del_directory:handle_call/3 - enqueue	149	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[I]	bodies@stor05.selectel.cloud.lan	2017-10-24 22:03:54.922410 +0300	1508871834	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```

On stor06:
```
[E]	bodies@stor06.selectel.cloud.lan	2017-10-24 22:03:51.610188 +0300	1508871831	leo_storage_handler_object:put/4	424	[{from,storage},{method,delete},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>},{req_id,14911882},{cause,not_found}]
[I]	bodies@stor06.selectel.cloud.lan	2017-10-24 22:03:51.612637 +0300	1508871831	leo_storage_handler_del_directory:handle_call/3 - enqueue	149	[{"msg: enqueued",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
[E]	bodies@stor06.selectel.cloud.lan	2017-10-24 22:03:52.710838 +0300	1508871832	leo_storage_handler_object:put/4	424	[{from,storage},{method,delete},{key,<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>},{req_id,73489324},{cause,not_found}]
[I]	bodies@stor06.selectel.cloud.lan	2017-10-24 22:03:55.66108 +0300	1508871835	leo_storage_handler_del_directory:run/5	731	[{"msg: dequeued and removed (dir)",<<"body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n">>}]
```

Whereis of other objects mentioned in these logs:
```
[vm@bodies-master ~]$ leofs-adm whereis 'body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n06d9c70d867d6ba1b14cbad1f4df775d'
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
 del?  |                  node                 |             ring address             |    size    |   checksum   |  has children  |  total chunks  |     clock      |             when            
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
  *    | bodies@stor05.selectel.cloud.lan      | 4a5dc2ff094f233b999f2ba45fd19f76     |         0B |   d41d8cd98f | false          |              0 | 55c4f9987284c  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor03.selectel.cloud.lan      | 4a5dc2ff094f233b999f2ba45fd19f76     |         0B |   d41d8cd98f | false          |              0 | 55c4f9987284c  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor01.selectel.cloud.lan      | 4a5dc2ff094f233b999f2ba45fd19f76     |         0B |   d41d8cd98f | false          |              0 | 55c4f9987284c  | 2017-10-24 22:04:25 +0300

[vm@bodies-master ~]$ leofs-adm whereis 'body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n'
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
 del?  |                  node                 |             ring address             |    size    |   checksum   |  has children  |  total chunks  |     clock      |             when            
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
       | bodies@stor03.selectel.cloud.lan      |                                      |            |              |                |                |                | 
       | bodies@stor02.selectel.cloud.lan      |                                      |            |              |                |                |                | 
       | bodies@stor06.selectel.cloud.lan      |                                      |            |              |                |                |                | 

[vm@bodies-master ~]$ leofs-adm whereis 'body/d5/08/87/d508877aec031efc066061366e5fc7d7ccf76c2e5bab2a8f4abfe0be03417e8d97a010c6cf81e241d6da1e42b2a04d98c023780000000000.xz\n44688f514c7dc8ec666e5fd3b18879db'
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
 del?  |                  node                 |             ring address             |    size    |   checksum   |  has children  |  total chunks  |     clock      |             when            
-------+---------------------------------------+--------------------------------------+------------+--------------+----------------+----------------+----------------+----------------------------
  *    | bodies@stor01.selectel.cloud.lan      | 93faa31df807feccce91196072afe50e     |         0B |   d41d8cd98f | false          |              0 | 55c4f997d28cb  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor03.selectel.cloud.lan      | 93faa31df807feccce91196072afe50e     |         0B |   d41d8cd98f | false          |              0 | 55c4f997d28cb  | 2017-10-24 22:04:25 +0300
  *    | bodies@stor02.selectel.cloud.lan      | 93faa31df807feccce91196072afe50e     |         0B |   d41d8cd98f | false          |              0 | 55c4f997d28cb  | 2017-10-24 22:04:25 +0300
```

Unfortunately, reducing amount of applications that in theory can upload the same object can't be done, we need that many uploaders. Avoiding upload of the same objects by different instances is impossible as well (introducing locking/skipping won't work because of multiple different destination objects in the same source object and already present locks for not processing the same source objects at once). Such conflicts occur not that often, actually (and it only can happen once for the same object). Besides these limitations, application logic can be changed but we need to be 100% sure that all objects are uploaded to storage and marked as present, the upload can be retried if needed but the object must never be lost and never be marked as present until it's actually was uploaded correctly. This problem is actually pretty critical, the only reason we didn't lose data here is because right now objects are actually uploaded on backup storage as well as LeoFS (as part of migration process) but it's only temporary as it has to work correctly with just LeoFS. But, like I said, it doesn't have to be changed in LeoFS itself as long as I can find a way to change application to correctly work around this problem. I should also note that the load on gateways and storage wasn't high so the reason why I got "service unavailable" aren't clear, they might have to do something with completing few MU requests for the same file at once and not the load itself.

**Comment by mocchira** [comment]
@vstax Thanks for reporting. now still WIP (how to fix or mitigate this problem ) however apparently the current error handling on leo_gateway against an objects multipart-uploaded can cause the object to be deleted unintentionally from application's point of view under many conflicts happened at once. we are considering how we can fix from LeoFS side.

> This problem is actually pretty critical, the only reason we didn't lose data here is because right now objects are actually uploaded on backup storage as well as LeoFS (as part of migration process) but it's only temporary as it has to work correctly with just LeoFS

For the record, if such data lost happened then there is the last resort (if compaction was still not triggered yet) retrieving raw data from AVS based on the result (offset and length for the target object) of diagnose. It maybe worth providing a tool to recover deleted objects for operational mistakes or this kind of events. 

> I should also note that the load on gateways and storage wasn't high so the reason why I got "service unavailable" aren't clear, they might have to do something with completing few MU requests for the same file at once and not the load itself.

OK. I will check if there is something on completing MU requests that can cause high load or 503 unintentionally.

**Comment by mocchira** [comment]
@vstax 

> OK. I will check if there is something on completing MU requests that can cause high load or 503 unintentionally.

Not on completing but I found things that can cause high load/503 on aborting MU so filed as another issue here: https://github.com/leo-project/leofs/issues/904

> now still WIP (how to fix or mitigate this problem ) however apparently the current error handling on leo_gateway against an objects multipart-uploaded can cause the object to be deleted unintentionally from application's point of view under many conflicts happened at once. we are considering how we can fix from LeoFS side

Possible fix on LeoFS side might be to check if there is already a parent record for the object when proceeding aborting MU request. if there is then we can think of the object as being created properly by some other uploader in parallel and don't remove the object and related ones (children/grandchildren).  Actually this is not perfect  as we have the issue https://github.com/leo-project/leofs/issues/719 however in your case, the every object having the same key being uploaded from many uploaders is guaranteed to be same then you will not suffer from #719 so this can be perfect fix for you.

**Comment by mocchira** [comment]
@vstax 

> Possible fix on LeoFS side might be to check if there is already a parent record for the object when proceeding aborting MU request. if there is then we can think of the object as being created properly by some other uploader in parallel and don't remove the object and related ones (children/grandchildren). Actually this is not perfect as we have the issue #719 however in your case, the every object having the same key being uploaded from many uploaders is guaranteed to be same then you will not suffer from #719 so this can be perfect fix for you

After discussing among the team, we've reached the consensus not to adopt the way above because this spec can cause the object multipart-uploaded then aborted to be immature (partially updated) in other cases (objects can be overwritten by a different content with the same key). enough documentations could mitigate the risk of this drawbacks but still complicated and easy to mess things up.

Only robust one is to fix #719 (means we get to have the object versioning) and remove only the objects without the latest stable version on aborting MU. however solving #719 must take much time so another solution would be needed.

Now we are considering to provide (a part of) conditional PUT in LeoFS that enable users to create the object only if it still doesn't exist (implement If-Non-Match: *). it seems to be relatively easy to implement on LeoFS for both (single part/multipart uploading) cases but might be difficult on the client side if the SDK doesn't provide I/F to set arbitrary HTTP headers when you call put or maltipart upload through aws-sdk. I'm going to check how boto3 handle this however let me know if you already know that also give us feedback about providing a part of conditional PUT in LeoFS would make sense to you.

**Comment by vstax** [comment]
Thanks for looking. I dont't know right away but we'll find a way to add
headers, we are using own wrapper around boto3 anyway. Buy that header will
be sent on all operations then, I think? From regular PUT to complete and
abort MU. Regular PUT will skip write operation then (OK), complete MU will
just delete temporary object (if it finds that object with that name
exists), and abort MU (if object exist) will just delete temp object as
well, right?

What about the race between finalizing other upload by other gateway and
head() on this one? If head didnt't see the object yet, won't gateway
proceed with real abort MU operation instead of just wiping temp object?

Another moment I don't quite understand is how using the header always to
turn on this logic all the time avoids the problem you described and want
to avoid (abort in case of immature data).

ср, 25 окт. 2017 г., 9:51 yoshiyuki.kanno <notifications@github.com>:

**Comment by vstax** [comment]
> Not on completing but I found things that can cause high load/503 on aborting MU so filed as another issue here: #904

I guess that's what these "leo_storage_handler_del_directory:run" calls come from? However, I don't think it fully explains where first of these aborts came from in the first place. These two abort MU errors are the only errors in application log at all (nothing else before and later; actually, no other errors since launch and till now at all). It feels like completing 5 MU operations at about the same time (all different clients and through 3 gateways) was what caused one of the aborts in the first place, is that possible? I can't tell the time of "complete MU" operation but I know that all 5 instances have started uploading this object within 0.3 second interval.

> For the record, if such data lost happened then there is the last resort (if compaction was still not triggered yet) retrieving raw data from AVS based on the result (offset and length for the target object) of diagnose. It maybe worth providing a tool to recover deleted objects for operational mistakes or this kind of events.

Ah yes sorry about strong words about "losing data", actually I was kind of thinking about something else. Yes, I keep in mind the ability to recover from AVS file, that simple format is really great since it makes pretty obvious how to do the recovery manually. I actually wanted to write a proper tool but failed to find a way to open leveldb with metadata, the (non-Erlang) implementations that I've tried - well, wrappers around C library generally - fail to open this database for one reason or another (they don't like outdated file extensions, generate different errors when files are renamed and so on) and since I didn't work with leveldb before I couldn't figure out how to make it work. Regarding scanning AVS file directly without paying attention to metadata, that's much simpler but apparently someone wrote a tool that goes through AVS already (seen it in main repo or somewhere around), around 2 pages of code in ruby, though I didn't check if it works with current version of AVS format. Shouldn't be too hard to modify it.

As for what I was thinking about - this file was easy to catch thanks to logs, but can I be sure that nothing of sorts can happen without any errors in storage logs? Our applications generally try to retrieve the object very soon after it's stored - while they might be still in gateway caches, so there will be no errors there as well (although if they fail to retrieve the object, we might notice that it's missing eventually). Logs from applications that upload objects can't be relied upon because they are designed to fail-restart-retry and single errors are expected from time to time.
Yes, we plan to do regular diagnose-start + comparing it to database (sometime later,  after migration is over) but it's not something that can be done often. So I wondered if there can be other cases similar to this which don't produce errors in logs of storage nodes.

**Comment by vstax** [comment]
@mocchira 
I'm able to add headers to boto3 request like this:
```
def inject_headers(params, **kwargs):
    params['headers']['If-None-Match'] = '*'

s3client = boto3.client(service_name='s3', ..... )
s3client.meta.events.register('before-call.s3', inject_headers)
```

The header is then used for all requests (CreateMultipartUpload/UploadPart/CompleteMultipartUpload)

**Comment by mocchira** [comment]
@vstax 

> EDIT: I think I understand now, you meant that you don't want to set a temporary logic that will be forced to change once #719 is implemented.

Probably Yes. I will upload the design doc on how the sequence between client/gateway/storage looks like when handling MPU with If-Non-Match: * so please give me FB.

> I guess that's what these "leo_storage_handler_del_directory:run" calls come from? 

Yes.

> However, I don't think it fully explains where first of these aborts came from in the first place. These two abort MU errors are the only errors in application log at all (nothing else before and later; actually, no other errors since launch and till now at all). It feels like completing 5 MU operations at about the same time (all different clients and through 3 gateways) was what caused one of the aborts in the first place, is that possible? I can't tell the time of "complete MU" operation but I know that all 5 instances have started uploading this object within 0.3 second interval.

Right. the reason of the first error is something different and

> It feels like completing 5 MU operations at about the same time (all different clients and through 3 gateways) was what caused one of the aborts in the first place, is that possible?

I think possible as each MU operation (6.2MB) cause
- 4 PUT
  - 1 for putting the temporary suffixed by UploadID on initiating
  - 2 for putting chunks on uploading parts
  - 1 for putting the parent on completing
- 5 HEAD
  - 2 for checking if the temporary really exists on uploading parts
  - 2 for retrieving each chunk and calculating md5 to validate the final shape on completing
  - 1 for retrieving one chunk to get the chunk size used in this MU
- 1 GET
  - 1 for getting the temporary on completing
- 1 DELETE
  - 1 for deleting the temporary on completing

on same storage nodes then result in some watchdog got triggered on one of them (I've found there are a few places leo_watchdog don't produce error lines even when it's fired. this should be fixed as another issue)

> Ah yes sorry about strong words about "losing data", actually I was kind of thinking about something else. Yes, I keep in mind the ability to recover from AVS file, that simple format is really great since it makes pretty obvious how to do the recovery manually. I actually wanted to write a proper tool but failed to find a way to open leveldb with metadata, the (non-Erlang) implementations that I've tried - well, wrappers around C library generally - fail to open this database for one reason or another (they don't like outdated file extensions, generate different errors when files are renamed and so on) and since I didn't work with leveldb before I couldn't figure out how to make it work. Regarding scanning AVS file directly without paying attention to metadata, that's much simpler but apparently someone wrote a tool that goes through AVS already (seen it in main repo or somewhere around), around 2 pages of code in ruby, though I didn't check if it works with current version of AVS format. Shouldn't be too hard to modify it.

maybe this one? https://github.com/leo-project/leofs_utils/blob/develop/tools/listup_avs.rb
(and just in case yes you don't need to refer leveldb to recover files deleted, only scanning AVS should work)
IIRC, this tool was written to detect garbage blocks that happen to be produced during compaction so the feature is different but the basic scanning logic could be reused.
Two thing to be paid attension
- Scanning part of this script can be reused if you don't use the custom metadata field (if you use, a little bit tweak needed)
- You need to find every chunk and concat those to recover the whole large object

We really appreciate if you would built the tool by yourself however don't hesitate to ask us to provide the official tool as I think it would be great if we provide the tool through leofs-adm like
```bash
$ leofs-adm recover-deleted-file bucket/path_to_file ${OUTPUT_DIR} ## gather every possible chunk and output every version already deleted under ${OUTPUT_DIR}
$ ls ${OUTPUT_DIR}
bucket/path_to_file_${timestamp}
```

> As for what I was thinking about - this file was easy to catch thanks to logs, but can I be sure that nothing of sorts can happen without any errors in storage logs? Our applications generally try to retrieve the object very soon after it's stored - while they might be still in gateway caches, so there will be no errors there as well (although if they fail to retrieve the object, we might notice that it's missing eventually). Logs from applications that upload objects can't be relied upon because they are designed to fail-restart-retry and single errors are expected from time to time.
Yes, we plan to do regular diagnose-start + comparing it to database (sometime later, after migration is over) but it's not something that can be done often. So I wondered if there can be other cases similar to this which don't produce errors in logs of storage nodes.

Nothing.
Only thing you'd have to check is "leo_storage_handler_del_directory:run" line that get appeared when abort MU requested in your case (no delete requests from applications).

> The header is then used for all requests (CreateMultipartUpload/UploadPart/CompleteMultipartUpload)

Thanks!

anyway, I just noticed is it possible to use not multipart-upload but singlepart one(normal put) from your applications? if possible, this could be short-term fix as unexpected deletes triggered by abort MU never happen.

**Comment by vstax** [comment]
@mocchira 
> on same storage nodes then result in some watchdog got triggered on one of them (I've found there are a few places leo_watchdog don't produce error lines even when it's fired. this should be fixed as another issue)

Just in case - on storage nodes rex and error watchdogs are enabled, cpu/disk/cluster are default (disabled).

> maybe this one?

Yes that's the one

> anyway, I just noticed is it possible to use not multipart-upload but singlepart one(normal put) from your applications? if possible, this could be short-term fix as unexpected deletes triggered by abort MU never happen.

IIRC the original reason we switched to S3 from REST (and then to boto3 right away because multipart uploads were a hassle in boto2) is that our experiments have shown pretty high load on gateways for uploading large objects in our scenarios (seems that it stores whole object in memory during upload and then splits them to parts inside), moving creation of multiple parts to clients solves this. I think it's a bit too late to switch to non-MU now that current load + migration (copying objects from old storages) is going on; however, during migration process - which should last almost a month - we are writing all the new data into LeoFS + backup place so as long as there is working workaround while we're still in migration phase we'll be able to recover any data deleted on LeoFS from backup anyway so we won't really lose anything.

**Comment by mocchira** [comment]
@vstax 

> Just in case - on storage nodes rex and error watchdogs are enabled, cpu/disk/cluster are default (disabled).

Thanks. now storage own watchdogs (restricting reqs based on the number of slow processing/timeout)  doesn't dump log lines and this watchdog currently always enabled! cause many timeouts could trigger on your env behind the scene. this should be fixed to allow users to enable/disable this watchdog and also dump log lines when it's fired.

> IIRC the original reason we switched to S3 from REST (and then to boto3 right away because multipart uploads were a hassle in boto2) is that our experiments have shown pretty high load on gateways for uploading large objects in our scenarios (seems that it stores whole object in memory during upload and then splits them to parts inside), moving creation of multiple parts to clients solves this. I think it's a bit too late to switch to non-MU now that current load + migration (copying objects from old storages) is going on; however, during migration process - which should last almost a month - we are writing all the new data into LeoFS + backup place so as long as there is working workaround while we're still in migration phase we'll be able to recover any data deleted on LeoFS from backup anyway so we won't really lose anything.

OK. I will proceed the conditional PUT.

**Comment by mocchira** [comment]
> Thanks. now storage own watchdogs (restricting reqs based on the number of slow processing/timeout) doesn't dump log lines and this watchdog currently always enabled! cause many timeouts could trigger on your env behind the scene. this should be fixed to allow users to enable/disable this watchdog and also dump log lines when it's fired.

filed on https://github.com/leo-project/leofs/issues/912

**Comment by vstax** [comment]
I had few more cases of this, interesting part is that in two of them I also got error messages on gateways (timestamps of these errors is 0.2-0.4s later than timestamps of "leo_storage_handler_del_directory:run/5" messages on storage nodes):
```
[E]     gateway@k01.selectel.cloud.lan  2017-10-28 08:17:43.39418 +0300 1509167863      leo_gateway_s3_api:handle_multi_upload_2/5      1282    [{key,"body/40/e8/13/40e813c425c7e960b8d0b6d29787f551234e927fc2767055c2b25085765eca6b8b87f1fc4ac08b894f603cac2848210b00a5b10000000000.xz"},{cause,invalid_metadata}]

[E]     gateway@k02.selectel.cloud.lan  2017-10-29 19:00:35.795675 +0300        1509292835      leo_gateway_s3_api:handle_multi_upload_2/5      1282    [{key,"body/0f/f0/68/0ff068a8a919bd240b4a95461b928eb02383a5deb2e7774bbdaf4bd6c7b614e032c61a393375e33cc57409f2977312af0042130100000000.xz"},{cause,invalid_metadata}]
```

**Comment by mocchira** [comment]
@vstax Thanks for sharing. This invalid_metadata was probably caused by not_found error as leo_storage_handler_del_directory:run got executed by other application. anyway the error message looks not intuitive so I will fix it to output the actual reason to fail.

As for conditional PUT, after some experiments, it turned out that this impl causes leo_storage to become high load especially in case num_of_containers is set to large number so we are considering other approach now such as 
- make it configurable whether deleting all related objects when receiving abort MU request or not through leo_gateway.conf (default: false). setting this to true would solve your problem.

**Comment by vstax** [comment]
@mocchira So with this option enabled, all abort MU operation will do is check for existence of temporary object and remove it if it exists - and nothing else, right? Regardless of presence of anything else?

This should work, I suppose. Maybe it'll be best to make it a hidden option so it can be silently removed once #719 is implemented?

**Comment by mocchira** [comment]
TODO
- [ ] Consider other possible solutions
- [ ] IMPL do nothing but removing a temporary object when handling Abort MU with its config enabled
- [ ] Fix the error log message - invalid_metadata

**Comment by mocchira** [comment]
Temporary fix https://github.com/leo-project/leofs/pull/919 got merged into develop.

**Comment by vstax** [comment]
@mocchira Thank you. I've upgraded and activated the option, I'll be able to tell how it goes in a few days (probably on Monday).

**Comment by vstax** [comment]
@mocchira 
The fix works, I do not get more deleted files anymore (or "leo_storage_handler_del_directory" messages). The only thing to note is that "get" errors for temporary objects are still appearing in same amounts. Often multiple times for the same temporary suffix (but sometimes 2-3 variants of suffixes for the same file)

```
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:32.137917 +0300	1509960812	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\n158a86aa5737a6e233dd36d574fc1ace">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:32.758753 +0300	1509960812	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\n158a86aa5737a6e233dd36d574fc1ace">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:44.609470 +0300	1509960824	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:44.711302 +0300	1509960824	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:44.932769 +0300	1509960824	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:45.122416 +0300	1509960825	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:45.595159 +0300	1509960825	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-06 12:33:45.876458 +0300	1509960825	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/f6/9a/83/f69a838564be44d11d554c21974b9bb3d0421e27584030dd33adeeedbbc3863122868edcbd6207a9cc5b1e5ea3426ef85027b20000000000.xz\nbbb57ba737454c72b6afbd6ccd2fd0bb">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-07 12:42:03.762880 +0300	1510047723	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n64957a18a9b72f70c5e478b72c5f531f">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-07 12:42:04.265142 +0300	1510047724	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n64957a18a9b72f70c5e478b72c5f531f">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-07 12:42:05.654110 +0300	1510047725	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n9400d0d8aea061c37d64cf380bd99335">>},{cause,not_found}]
[E]	bodies@stor01.selectel.cloud.lan	2017-11-07 12:42:06.652033 +0300	1510047726	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n9400d0d8aea061c37d64cf380bd99335">>},{cause,not_found}]
```
Besides the fact that these errors still appear (shouldn't they be gone since now all that should happen to temporary object is that it's created once and removed once during either Abort MU or Complete MU operation?), what alarms me is that the name of temporary object is the same. Supposedly I'm uploading object with the same name from multiple clients at once.. Shouldn't each get its own random suffix? Right now it just feels as if multiple clients got the same suffix, is this possible? Or I'm misinterpreting these messages?

For example, the last object above was uploaded (and downloaded, as each upload triggers at least one download) multiple times. Of course, it appears in logs of other nodes with different suffixes because location of temporary object depends on its name:
```
[E]	bodies@stor02.selectel.cloud.lan	2017-11-07 12:42:03.761494 +0300	1510047723	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n64957a18a9b72f70c5e478b72c5f531f">>},{cause,not_found}]
[E]	bodies@stor02.selectel.cloud.lan	2017-11-07 12:42:04.264094 +0300	1510047724	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n64957a18a9b72f70c5e478b72c5f531f">>},{cause,not_found}]
[E]	bodies@stor02.selectel.cloud.lan	2017-11-07 12:42:09.116867 +0300	1510047729	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n050a2fb91077fddc080f8428088bf1b4">>},{cause,not_found}]
[E]	bodies@stor02.selectel.cloud.lan	2017-11-07 12:42:09.939177 +0300	1510047729	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n050a2fb91077fddc080f8428088bf1b4">>},{cause,not_found}]
```
```
[E]	bodies@stor03.selectel.cloud.lan	2017-11-07 12:42:09.116037 +0300	1510047729	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n050a2fb91077fddc080f8428088bf1b4">>},{cause,not_found}]
[E]	bodies@stor03.selectel.cloud.lan	2017-11-07 12:42:09.938490 +0300	1510047729	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n050a2fb91077fddc080f8428088bf1b4">>},{cause,not_found}]
```
```
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:08.486035 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:08.732659 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:08.773555 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:08.802392 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:08.868787 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:09.87257 +0300	1510047729	leo_storage_handler_object:get/190	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor04.selectel.cloud.lan	2017-11-07 12:42:09.94030 +0300	1510047729	leo_storage_handler_object:get/190	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
```
```
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:08.368011 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:08.646328 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:08.772892 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:08.801757 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:08.868075 +0300	1510047728	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:09.86535 +0300	1510047729	leo_storage_handler_object:get/190	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:09.93226 +0300	1510047729	leo_storage_handler_object:get/190	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
[E]	bodies@stor05.selectel.cloud.lan	2017-11-07 12:42:09.300244 +0300	1510047729	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\ne4abe712d9c0aac8b05615233781d3e3">>},{cause,not_found}]
```
```
[E]	bodies@stor06.selectel.cloud.lan	2017-11-07 12:42:05.654821 +0300	1510047725	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n9400d0d8aea061c37d64cf380bd99335">>},{cause,not_found}]
[E]	bodies@stor06.selectel.cloud.lan	2017-11-07 12:42:06.662987 +0300	1510047726	leo_storage_handler_object:get/1	90	[{from,storage},{method,get},{key,<<"body/75/21/8b/75218be76f7046f2aa1fadfc4a6ffef16aa24c425862d88dae9153759fca070a642eda74fa3e8a3568571c8e2f565bc2e879b20000000000.xz\n9400d0d8aea061c37d64cf380bd99335">>},{cause,not_found}]
```

"Whereis" for these objects show that they were deleted slightly before the corresponding log message. I don't understand why temporary objects are being deleted so many times... Another interesting fact is that I see only 5 different suffixes here, but client shows 7 errors during CompleteMultipartUpload (each should've been causing an Abort, as I understand) - these errors have happened from 12:42:04.307 to 12:42:09.958. Maybe the other 2 Abort MU operations have just removed temporary object without any problems so there are no errors in logs.

**Comment by mocchira** [comment]
@vstax 

> The fix works, I do not get more deleted files anymore (or "leo_storage_handler_del_directory" messages). 

Great to hear.

> Supposedly I'm uploading object with the same name from multiple clients at once.. Shouldn't each get its own random suffix? Right now it just feels as if multiple clients got the same suffix, is this possible? Or I'm misinterpreting these messages?

Yep, good catch! It turned out that now the upload_id can be conflicted when the multiple clients try to upload a same object at once. I'm going to fix by generating upload_id based on UUIDv4 instead.

**Comment by vstax** [comment]
@mocchira 

>  It turned out that now the upload_id can be conflicted when the multiple clients try to upload a same object at once

Ooh, that would explain why aborts started to happened in the first place. I bet they would've completely gone (without the need for change from this ticket) with unique upload_id's.

**Comment by vstax** [comment]
Upgraded to the latest version with upload_id fix, I'll be able to tell if there are any aborts at all in a few days.

**Comment by vstax** [comment]
@mocchira With upload_id fix, there are no more aborts on client - multiple MU complete without problems now (and no errors in storage logs). Thank you; there seem to be nothing else about this issue so closing.
